File: go/pages.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"strconv"
	"strings"

	"golang.org/x/net/html"
)

var domain = "https://www.divan.ru"

type divanData struct {
	url, name, sale, art string
	price, oldPrice      int
}

func getDivans() []divanData {
	var divans []divanData

	paths := []string{
		"/category/stok-mebeli?categories[]=2",
		"/category/stok-mebeli/page-2?categories[]=2",
		"/category/stok-mebeli/page-3?categories[]=2",
	}
	for _, path := range paths {
		res, err := getHtml(path)
		if err != nil {
			fmt.Fprintf(os.Stderr, "parse: %v\n", err)
		}
		divans = append(divans, res...)
	}
	return divans
}

func getHtml(path string) ([]divanData, error) {
	url := domain + path
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode != http.StatusOK {
		resp.Body.Close()
		return nil, fmt.Errorf("getting %s: %s", url, resp.Status)
	}
	doc, err := html.Parse(resp.Body)
	resp.Body.Close()
	if err != nil {
		return nil, fmt.Errorf("пирсинг %s: as HTML: %v", url, err)
	}
	divansHtml := getDivansHtml(nil, doc)

	return getDivansDataFromHtml(divansHtml), nil

}

func getDivansHtml(items []*html.Node, n *html.Node) []*html.Node {
	if n.Type == html.ElementNode {
		for _, a := range n.Attr {
			if a.Key == "class" && strings.Contains(a.Val, "lsooF") {
				items = append(items, n)
			}
		}
	}

	for c := n.FirstChild; c != nil; c = c.NextSibling {
		items = getDivansHtml(items, c)
	}
	return items
}

func getDivansDataFromHtml(items []*html.Node) []divanData {
	var divans []divanData
	for _, divanHtml := range items {
		data := getDivanDataFromHtml(divanHtml)

		divans = append(divans, data)

	}
	return divans
}

func getDivanDataFromHtml(divanHtml *html.Node) divanData {
	var data divanData

	for _, a := range divanHtml.FirstChild.Attr {
		if a.Key == "href" {
			data.url = domain + a.Val
			_, art, found := strings.Cut(a.Val, "art--")
			if found {
				data.art = art
			}
		}
	}

	data.name = divanHtml.FirstChild.FirstChild.Data
	thirdChild := divanHtml.FirstChild.NextSibling.NextSibling

	if thirdChild == nil {
		data.price = 0
		data.oldPrice = 0
		data.sale = "0%"
	} else {
		data.price = getPrice(thirdChild.FirstChild.FirstChild.Data)
		saleNode := thirdChild.FirstChild.NextSibling
		if saleNode == nil {
			data.oldPrice = 0
			data.sale = "0%"
		} else {
			data.oldPrice = getPrice(saleNode.FirstChild.Data)
			data.sale = saleNode.NextSibling.FirstChild.FirstChild.Data
		}
	}

	return data
}

func getPrice(n string) int {

	nTrim := strings.ReplaceAll(n, " ", "")
	res, err := strconv.Atoi(nTrim)
	if err == nil {
		return res
	} else {
		fmt.Println(err)
		return 0
	}

}

File: pages.js
const axios = require('axios').default;
const { parse } = require('node-html-parser');
const _ = require('lodash');
const fs = require('fs');

const domain = 'https://www.divan.ru';

const getUrl = (path) => `${domain}${path}`

async function getPageItems(path, selector) {
    const { data } = await axios.get(getUrl(path))
    return parse(data).querySelectorAll(selector);
}

async function getPagesPath() {
    const res = await getPageItems('/category/stok-mebeli?categories[]=2', '.dqBvL');
    const [{ childNodes }] = res;
    const pages = childNodes.reduce((t, item) => {
        const href = _.get(item, 'attrs.href') || '';
        if (href.includes('/category/stok-mebeli') && !t.includes(href))
            t.push(href)
        return t

    }, [])

    return pages

}


function getData(element) {
    const name = _.get(element, 'childNodes[0].childNodes[0].text')?.trim();
    const href = _.get(element, 'childNodes[0].attrs.href');
    const price = _.get(element, 'childNodes[2].childNodes[0].childNodes[0].text')?.trim();
    const oldPrice = _.get(element, 'childNodes[2].childNodes[1].childNodes[0].text')?.trim();
    const sale = _.get(element, 'childNodes[2].childNodes[2].childNodes[0].childNodes[0].text')?.trim();

    const url = getUrl(href);
    const atr = href.split('art--').pop()

    return {
        name, sale, url, atr,
        price: Number(price?.replace(' ', '') || 0),
        oldPrice: Number(oldPrice?.replace(' ', '') || 0),
        date: new Date(),
    }
}


async function getNewData() {
    // const pagesPaths = await getPagesPath();

    const pagesPaths = [
        '/category/stok-mebeli?categories[]=2',
        '/category/stok-mebeli/page-2?categories[]=2',
        '/category/stok-mebeli/page-3?categories[]=2'
    ];

    const items = [];

    for (const path of pagesPaths) {
        console.log('page', path);
        const pageItems = await getPageItems(path, '.lsooF')
        items.push(...pageItems)
    }

    return items.map(getData);

}

async function writeDB() {
    const data = await getNewData();
    console.log(data);
    fs.writeFileSync('./db.json', JSON.stringify(data, null, 2))


}

module.exports = {
    getNewData,
    writeDB
}